File: leetcode/solved/math/entropy.java
class Solution {
    public double calculateEntropy(int[] input) {
        Map<Integer, Double> P = new HashMap<> ();
        
        for (int in : input) {
            P.put(in, P.getOrDefault(in, 0.) + 1.);
        }
        
        double total = input.length, s = 0;
        
        Set<Integer> values = P.keySet ();
        for (Integer k : values) {
            double pi = P.get(k) / total;
            s += pi * Math.log(pi) / Math.log(2.);
        }
        
        // fix : rehefa C++
        // double d = Math.abs(s);
        // if (d <= 10e-6) return 0;
        return -s;
    }
}

File: leetcode/solved/math/entropy.cpp
class Solution {
public:
    double calculateEntropy(vector<int>& input) {
        unordered_map<int, double> P;
        
        for (int &in : input) P[in]++;
        
        double total = input.size(), s = 0;
        for (auto p : P) {
            double pi = p.second / total;
            s += pi * log(pi) / log(2.);
        }
        
        double d = fabs(s);
        if (d <= 10e-6) return 0;
        return -s;
    }
};